Refactor JSC_API_AVAILABLE to behave more like WK_API_AVAILABLE

emw-apple opened the pull request:
#### a0da8012d30d24e63c583aa54056f2dedcdad4ac
<pre>
Refactor JSC_API_AVAILABLE to behave more like WK_API_AVAILABLE
<a href="https://bugs.webkit.org/show_bug.cgi?id=250740">https://bugs.webkit.org/show_bug.cgi?id=250740</a>
rdar://104354248

Reviewed by Alexey Proskuryakov.

Previously, we did not turn off JSC&apos;s availability declarations in
downlevel builds. This prevented us from converting TBA declarations to
shipping versions until those OS versions are used in all downlevel
builds. As a result, the most recent JSC_API_AVAILABLE versions are
macos(11.0) and ios(14.0).

Fix this and simplify the build by chaging JSC&apos;s availability logic to
be more like WK_API_AVAILABLE:

- In debug/release builds, JSC_API_AVAILABLE postprocessing is disabled.
- In production builds, WebKitAdditions determines when postprocessing
  can be disabled, supporting the use of more recent availability
  versions.
- When postprocessing is disabled, a public JSC header disables
  JSC_API_AVAILABLE via preprocessor macros.

Postprocessing can be disabled by the
JSC_FRAMEWORK_HEADER_POSTPROCESSING_DISABLED build setting, but
WK_FRAMEWORK_HEADER_POSTPROCESSING_DISABLED is also recognized to match
existing behavior in WebKitAdditions.

* Source/JavaScriptCore/API/JSRemoteInspector.h: Remove the workaround
  Patrick Angle needed to be able to call new SPI from downlevel builds.

* Source/JavaScriptCore/API/WebKitAvailability.h: Behave like
  &lt;WebKit/WKFoundation.h&gt;. When JSC header postprocessing is disabled,
  macros are used to stub out the availiability declarations. This
  conditional logic is unifdef&apos;d appropriately during the build.

* Source/JavaScriptCore/API/tests/testIncludes.m: No longer need to stub
  out JSC_API_AVAILABLE here.

* Source/JavaScriptCore/Configurations/JavaScriptCore.xcconfig: Disable
  postprocessing in debug/release, downlevel, and STP builds, just like
  WebKit does.

* Source/JavaScriptCore/Scripts/postprocess-header-rule: Synced with
  WebKit&apos;s postprocess-header-rule. Largely equivalent,
  except with &quot;JSC&quot; prefixes. There is no needed equivalent to WebKit&apos;s
  replace-webkit-additions-includes.py call or most of its unifdef
  logic.
  JSC_MAC_VERSION_TBA and JSC_IOS_VERSION_TBA, which are TBA&apos;d
  versions of __MAC_OS_X_VERSION_MIN_REQUIRED and
  __IPHONE_OS_VERSION_MIN_REQUIRED, are still emitted, though we are not
  currently using them.

* Source/WebCore/config.h: No longer need to stub out JSC_API_AVAILABLE
  for CMake builds here, as JSC now does this for us.
* Source/WebKit/WebKit2Prefix.h: Ditto.
* Source/WebKitLegacy/mac/WebKitPrefix.h: Ditto.

Canonical link: <a href="https://commits.webkit.org/260542@main">https://commits.webkit.org/260542@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8e3e34d87bb271528e3768acc332c63c9c50b2b6

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/108466 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/17562 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/41316 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/117571 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/117785 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/112353 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/19013 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/8840 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/100692 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/114234 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/14249 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/97469 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/42210 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/96227 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/29112 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/83913 "Found 1 new API test failure: /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/text/state-changed (failure)") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/97620 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/10378 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/30458 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/98529 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/8498 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/11131 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/7372 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/30952 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/16525 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/50054 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/106042 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7288 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/12717 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/26246 "Passed tests") | 
| | | | | 
<!--EWS-Status-Bubble-End-->

emw-apple commented on Source/JavaScriptCore/Scripts/postprocess-header-rule:
```diff
@@ -91,14 +91,17 @@ function rewrite_headers () {
-}
-
-DEFINITIONS_PATH=usr/local/include/WebKitAdditions/Scripts/postprocess-framework-headers-definitions
+UNIFDEF_OPTIONS=(-Bt)
```
`-t` turns off some unneeded source parsing behavior. unifdef does not understand c++14 style single-quotes used as separators in integer literals, like
https://github.com/WebKit/WebKit/blob/f67bf0023bc8a15007490b887f367c001fbddf93/Source/JavaScriptCore/assembler/ARM64Assembler.h#L1600

aproskuryakov commented on Source/JavaScriptCore/API/JSRemoteInspector.h:
```diff
@@ -50,10 +50,8 @@ JS_EXPORT void JSRemoteInspectorSetLogToSystemConsole(bool) JSC_API_AVAILABLE(ma
-// downlevel builds. These SPI are available starting in JSC_MAC_TBA and JSC_IOS_TBA.
-JS_EXPORT bool JSRemoteInspectorGetInspectionFollowsInternalPolicies(void);
-JS_EXPORT void JSRemoteInspectorSetInspectionFollowsInternalPolicies(bool);
+JS_EXPORT bool JSRemoteInspectorGetInspectionFollowsInternalPolicies(void) JSC_API_AVAILABLE(macos(JSC_MAC_TBA), ios(JSC_IOS_TBA));
```
Is TBA still accurate for these now?

patrickangle replied:
It can be the versions we seeded to developers yesterday. But also fine if that is being done as a followup for all the applicable uses. Landing with the TBAs would be a way to ensure the fix works across our builders, though!

emw-apple replied:
That's my plan. I'll land with TBAs, and set this to the correct version as part of https://bugs.webkit.org/show_bug.cgi?id=252437 (Update WebKit TBA versions for macOS 13.3 and iOS 16.4)

aproskuryakov commented on Source/JavaScriptCore/Scripts/postprocess-header-rule:
```diff
@@ -43,11 +38,17 @@ function process_definitions () {
+
+process_definitions "${BUILT_PRODUCTS_DIR}/${DEFINITIONS_PATH}" || process_definitions "${SDKROOT}/${DEFINITIONS_PATH}"
+
+# For historical compatibility, JSC's postprocessing may be toggled separately
```
I'm not sure what this means. People certainly build JSC separately, but that's not historical, so I must be thinking in the wrong direction.

aproskuryakov reviewed (approved):
More red than green in the diff => ship it! :-)

emw-apple committed:
Refactor JSC_API_AVAILABLE to behave more like WK_API_AVAILABLE
https://bugs.webkit.org/show_bug.cgi?id=250740
rdar://104354248

Reviewed by Alexey Proskuryakov.

Previously, we did not turn off JSC's availability declarations in
downlevel builds. This prevented us from converting TBA declarations to
shipping versions until those OS versions are used in all downlevel
builds. As a result, the most recent JSC_API_AVAILABLE versions are
macos(11.0) and ios(14.0).

Fix this and simplify the build by chaging JSC's availability logic to
be more like WK_API_AVAILABLE:

- In debug/release builds, JSC_API_AVAILABLE postprocessing is disabled.
- In production builds, WebKitAdditions determines when postprocessing
  can be disabled, supporting the use of more recent availability
  versions.
- When postprocessing is disabled, a public JSC header disables
  JSC_API_AVAILABLE via preprocessor macros.

Postprocessing can be disabled by the
JSC_FRAMEWORK_HEADER_POSTPROCESSING_DISABLED build setting, but
WK_FRAMEWORK_HEADER_POSTPROCESSING_DISABLED is also recognized to match
existing behavior in WebKitAdditions.

* Source/JavaScriptCore/API/JSRemoteInspector.h: Remove the workaround
  Patrick Angle needed to be able to call new SPI from downlevel builds.

* Source/JavaScriptCore/API/WebKitAvailability.h: Behave like
  <WebKit/WKFoundation.h>. When JSC header postprocessing is disabled,
  macros are used to stub out the availiability declarations. This
  conditional logic is unifdef'd appropriately during the build.

* Source/JavaScriptCore/API/tests/testIncludes.m: No longer need to stub
  out JSC_API_AVAILABLE here.

* Source/JavaScriptCore/Configurations/JavaScriptCore.xcconfig: Disable
  postprocessing in debug/release, downlevel, and STP builds, just like
  WebKit does.

* Source/JavaScriptCore/Scripts/postprocess-header-rule: Synced with
  WebKit's postprocess-header-rule. Largely equivalent,
  except with "JSC" prefixes. There is no needed equivalent to WebKit's
  replace-webkit-additions-includes.py call or most of its unifdef
  logic.
  JSC_MAC_VERSION_TBA and JSC_IOS_VERSION_TBA, which are TBA'd
  versions of __MAC_OS_X_VERSION_MIN_REQUIRED and
  __IPHONE_OS_VERSION_MIN_REQUIRED, are still emitted, though we are not
  currently using them.

* Source/WebCore/config.h: No longer need to stub out JSC_API_AVAILABLE
  for CMake builds here, as JSC now does this for us.
* Source/WebKit/WebKit2Prefix.h: Ditto.
* Source/WebKitLegacy/mac/WebKitPrefix.h: Ditto.

Canonical link: https://commits.webkit.org/260542@main
